File: game_server/src/sc/server/Lobby.java
package sc.server;

import java.io.IOException;
import java.util.Iterator;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import sc.api.plugins.exceptions.RescueableClientException;
import sc.protocol.requests.AuthenticateRequest;
import sc.protocol.requests.CancelRequest;
import sc.protocol.requests.FreeReservationRequest;
import sc.protocol.requests.ILobbyRequest;
import sc.protocol.requests.JoinPreparedRoomRequest;
import sc.protocol.requests.JoinRoomRequest;
import sc.protocol.requests.ObservationRequest;
import sc.protocol.requests.PauseGameRequest;
import sc.protocol.requests.PrepareGameRequest;
import sc.protocol.requests.StepRequest;
import sc.protocol.responses.RoomPacket;
import sc.server.gaming.GameRoom;
import sc.server.gaming.GameRoomManager;
import sc.server.gaming.PlayerRole;
import sc.server.gaming.ReservationManager;
import sc.server.network.Client;
import sc.server.network.ClientManager;
import sc.server.network.IClientListener;
import sc.server.network.IClientRole;
import sc.server.network.PacketCallback;

/**
 * The lobby will help clients find a open game or create new games to play with
 * another client.
 * 
 * @author mja
 * @author rra
 */
public class Lobby implements IClientManagerListener, IClientListener
{
	private final Logger			logger			= LoggerFactory
															.getLogger(Lobby.class);
	private final GameRoomManager	gameManager		= new GameRoomManager();
	private final ClientManager		clientManager	= new ClientManager();

	public Lobby()
	{
		this.clientManager.addListener(this);
	}

	public void start() throws IOException
	{
		this.gameManager.start();
		this.clientManager.start();
	}

	@Override
	public void onClientConnected(Client client)
	{
		client.addClientListener(this);
		client.start();
	}

	@Override
	public void onClientDisconnected(Client source)
	{
		/*final Client client = source;
		new Thread(new Runnable() {
			@Override
			public void run()
			{
				synchronized(client) {
					client.removeClientListener(Lobby.this);
				}
			}
		}).start();*/
		this.logger.info("{} disconnected.", source);
		source.removeClientListener(this);
	}

	@Override
	public void onRequest(Client source, PacketCallback callback)
			throws RescueableClientException
	{
		Object packet = callback.getPacket();

		if (packet instanceof ILobbyRequest)
		{
			if (packet instanceof JoinPreparedRoomRequest)
			{
				ReservationManager
						.redeemReservationCode(source,
								((JoinPreparedRoomRequest) packet)
										.getReservationCode());

			}
			else if (packet instanceof JoinRoomRequest)
			{
				this.gameManager.joinOrCreateGame(source,
						((JoinRoomRequest) packet).getGameType());
			}
			else if (packet instanceof AuthenticateRequest)
			{
				source.authenticate(((AuthenticateRequest) packet)
						.getPassword());
			}
			else if (packet instanceof PrepareGameRequest)
			{
				PrepareGameRequest prepared = (PrepareGameRequest) packet;
				source.send(this.gameManager.prepareGame(prepared));
				/*source.send(this.gameManager.prepareGame(
						prepared.getGameType(), prepared.getPlayerCount(),
						prepared.getSlotDescriptors()));*/
			}
			else if (packet instanceof FreeReservationRequest)
			{
				FreeReservationRequest request = (FreeReservationRequest) packet;
				ReservationManager.freeReservation(request.getReservation());
			}
			else if (packet instanceof RoomPacket)	// e.g. new move
			{
				RoomPacket casted = (RoomPacket) packet;
				GameRoom room = this.gameManager.findRoom(casted.getRoomId());
				room.onEvent(source, casted.getData());
			}
			else if (packet instanceof ObservationRequest)
			{
				// TODO: check permissions
				ObservationRequest observe = (ObservationRequest) packet;
				GameRoom room = this.gameManager.findRoom(observe.getRoomId());
				room.addObserver(source);
			}
			else if (packet instanceof PauseGameRequest)
			{
				PauseGameRequest pause = (PauseGameRequest) packet;
				GameRoom room = this.gameManager.findRoom(pause.roomId);
				room.pause(pause.pause);
			}
			else if (packet instanceof StepRequest)
			{
				StepRequest pause = (StepRequest) packet;
				GameRoom room = this.gameManager.findRoom(pause.roomId);
				room.step(pause.forced);
			}
			else if (packet instanceof CancelRequest)
			{
				CancelRequest cancel = (CancelRequest) packet;
				GameRoom room = this.gameManager.findRoom(cancel.roomId);
				room.cancel();
			}
			else
			{
				throw new RescueableClientException(
						"Unhandled Packet of type: " + packet.getClass());
			}

			callback.setProcessed();
		}
	}

	public void close()
	{
		this.clientManager.close();
		this.gameManager.close();
	}

	public GameRoomManager getGameManager()
	{
		return this.gameManager;
	}

	public ClientManager getClientManager()
	{
		return this.clientManager;
	}

	@Override
	public void onError(Client source, Object errorPacket)
	{
		for (Iterator<IClientRole> iterator = source.getRoles().iterator(); iterator.hasNext();)
		{
			PlayerRole role = (PlayerRole) iterator.next();
			role.getPlayerSlot().getRoom().onClientError(source, errorPacket);
		}
	}
}

File: server/src/sc/server/Lobby.kt
package sc.server

import org.slf4j.LoggerFactory
import sc.api.plugins.exceptions.RescuableClientException
import sc.protocol.room.RoomPacket
import sc.protocol.requests.*
import sc.protocol.responses.*
import sc.protocol.room.ErrorMessage
import sc.server.gaming.GameRoomManager
import sc.server.gaming.PlayerRole
import sc.server.gaming.ReservationManager
import sc.server.network.*
import sc.shared.InvalidGameStateException
import sc.shared.Score
import java.io.Closeable
import java.io.IOException

/** The lobby joins clients into a game by finding open rooms or creating new ones. */
open class Lobby: GameRoomManager(), IClientListener, Closeable {
    private val logger = LoggerFactory.getLogger(Lobby::class.java)
    
    val clientManager = ClientManager().also {
        it.setOnClientConnected(this::onClientConnected)
    }
    
    /** @see ClientManager.start */
    @Throws(IOException::class)
    fun start() {
        clientManager.start()
    }
    
    /**
     * Add lobby as listener to client.
     * Prepare client for send and receive.
     *
     * @param client connected XStreamClient
     */
    fun onClientConnected(client: Client) {
        client.addClientListener(this)
        client.start()
    }
    
    override fun onClientDisconnected(source: Client) {
        logger.info("{} disconnected.", source)
        source.removeClientListener(this)
    }
    
    /** Handle requests or moves of clients. */
    @Throws(RescuableClientException::class, InvalidGameStateException::class)
    override fun onRequest(source: Client, callback: PacketCallback) {
        when (val packet = callback.packet) {
            is RoomPacket -> {
                // i.e. new move
                val room = this.findRoom(packet.roomId)
                room.onEvent(source, packet.data)
            }
            is JoinPreparedRoomRequest ->
                try {
                    ReservationManager.redeemReservationCode(source, packet.reservationCode)
                } catch (e: RescuableClientException) {
                    source.send(ErrorPacket(packet, e.message))
                }
            is JoinRoomRequest -> {
                val gameRoomMessage = this.joinOrCreateGame(source, packet.gameType)
                // null is returned if join was unsuccessful
                if (gameRoomMessage != null) {
                    clientManager.clients
                            .filter { it.isAdministrator }
                            .forEach { it.send(gameRoomMessage) }
                }
            }
            is AuthenticateRequest -> source.authenticate(packet.password)
            is AdminLobbyRequest -> {
                if (!source.isAdministrator)
                    throw UnauthenticatedException(packet)
                when (packet) {
                    is PrepareGameRequest -> {
                        source.send(this.prepareGame(packet))
                    }
                    is FreeReservationRequest -> {
                        ReservationManager.freeReservation(packet.reservation)
                    }
                    is ControlTimeoutRequest -> {
                        val room = this.findRoom(packet.roomId)
                        room.ensureOpenSlots(packet.slot + 1)
                        val slot = room.slots[packet.slot]
                        slot.descriptor = slot.descriptor.copy(canTimeout = packet.activate)
                        slot.role?.player?.canTimeout = packet.activate
                    }
                    is ObservationRequest -> {
                        val room = this.findRoom(packet.roomId)
                        room.addObserver(source)
                    }
                    is PauseGameRequest -> {
                        try {
                            val room = this.findRoom(packet.roomId)
                            room.pause(packet.pause)
                        } catch (e: RescuableClientException) {
                            this.logger.error("Got exception on pause: {}", e)
                        }
                    }
                    is StepRequest -> {
                        // TODO check for a prior pending StepRequest
                        val room = this.findRoom(packet.roomId)
                        room.step(packet.forced)
                    }
                    is CancelRequest -> {
                        requireNotNull(packet.roomId) { "Can't cancel a game with roomId null!" }
                        val room = this.findRoom(packet.roomId)
                        room.cancel()
                    }
                    is PlayerScoreRequest -> {
                        val displayName = packet.displayName
                        val score = getScoreOfPlayer(displayName)
                                    ?: throw IllegalArgumentException("Score for \"$displayName\" could not be found!")
                        logger.debug("Sending score of player \"{}\"", displayName)
                        source.send(PlayerScoreResponse(score))
                    }
                    is TestModeRequest -> {
                        val testMode = packet.testMode
                        logger.info("Setting Test mode to {}", testMode)
                        Configuration.set(Configuration.TEST_MODE, testMode.toString())
                        source.send(TestModeResponse(testMode))
                    }
                }
            }
            else -> throw RescuableClientException("Unhandled Packet of type: " + packet.javaClass)
        }
        callback.setProcessed()
    }
    
    private fun getScoreOfPlayer(displayName: String): Score? {
        for (score in this.scores) {
            if (score.displayName == displayName) {
                return score
            }
        }
        return null
    }
    
    override fun close() {
        clientManager.close()
    }
    
    override fun onError(source: Client, errorPacket: ErrorMessage) {
        source.roles.filterIsInstance<PlayerRole>().forEach {
            it.playerSlot.room.onClientError(errorPacket)
        }
    }
}
